### PR TITLE
Issue/2244 diff view

### LINKED ISSUE #2244: Design a diff functionality for desired state
The goal of this ticket is to design/mock/investigate a diff functionality that can compare two desired state versions. 

## Data
Attached to this issue is a set of scripts and data based on our 5g demo. 

The dl.py script downloads the desired state of all versions and resources and places this in the out directory (also in the zip). The diff.py script generates a json based diff between two versions of desired state.

The current proposed structure for the desired state is:
- A mapping of added resources
- A mapping of removed resources
- A mapping of modified resources

The mapping is resource id to attributes. For each attribute the r1 (source version of the diff) and r2 (target version of the diff) is provided + a text based diff.

The r1 and r2 data is always provided. If the resource is removed or added the r1 or r2 has null as value.

For the text based diff, the diff.py script tries to guess some of the types and then convert it to a string representation. For example a list is sorted and then rendered as string with an element per line. Json data is rendered to json, ...

File data is not yet fetched.

[diff.zip](https://github.com/inmanta/web-console/files/7769963/diff.zip)

The diff.json file is generated between version 4110 and 4122

## Visualisation

A good example of a well known UI for visualizing diffs is how for example gitlab (or github) does this. A view where on the side you have a list of all the changed resources using green and red to indicate add, delete or modify and then a diff overview per attribute per resource.

image

This might be a continuous list or per resource.

## Results

- Notes data format, libraries, ...
- Mocks
- Small MVP?

### PR DESCRIPTION
**Pull request by Stijnkool**
# Description

- Selecting 2 version and getting to the compare page is already implemented
- Storybook contains some examples
- I have added my analysis to the docs folder. This contains my thoughts.

closes #2244 

# Self Check:

Strike through any lines that are not applicable (`~~line~~`) then check the box

- [x] Attached issue to pull request
- [x] Changelog entry
- [x] Code is clear and sufficiently documented
- [x] Sufficient test cases (reproduces the bug/tests the requested feature)
- [x] Correct, in line with design
- [ ] ~~End user documentation is included or an issue is created for end-user documentation (add ref to issue here: )~~

# Reviewer Checklist:

- [ ] Sufficient test cases (reproduces the bug/tests the requested feature)
- [ ] Code is clear and sufficiently documented
- [ ] Correct, in line with design

### CODE REVIEW

File: docs/desired-state-diff/index.md [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+For the actuall diff visual, we have 2 options:
+
+### Option 1, build the visual ourselves
+
+We have complete freedom to do what we want. But we might not need it.  
+We can use `jsdiff` to generate a json structure describing the unchanged, deleted and added characters.
+
+### Option 2, use a lib for generating the visual
+
+We can use a component like `react-diff-viewer`. There are some options. But we will always be limited.
```
Comment (andraskvr) [inline docs/desired-state-diff/index.md:202]: For me, the question is that is there a library that's good enough for our purposes?

**Review by andraskvr — COMMENTED** [review]
Some questions and comments, but overall nice work on the docs

File: src/UI/Pages/DesiredState/Components/CompareAction.tsx [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+  const navigateTo = useNavigateTo();
+  const onClick = () => {
+    navigateTo("DesiredStateCompare", {
+      targetVersion: version.toString(),
+      sourceVersion: sourceVersion.toString(),
+    });
+  };
+
+  return (
+    <DropdownItem onClick={onClick} icon={<BalanceScaleIcon />}>
```
[the remark this answers is not in the document]
Comment (Stijnkool) [reply src/UI/Pages/DesiredState/Components/CompareAction.tsx:81]: Is this a patternfly or font awesome icon?

[the remark this answers is not in the document]
Comment (Stijnkool) [reply src/UI/Pages/DesiredState/Components/CompareAction.tsx:81]: If required, we can add a different icon later in the finalisation of this feature.

File: docs/desired-state-diff/index.md [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+    {
+      "resource_id": "kubernetes_calico::resources::BGPPeer[dc-1,identifier=/cluster/dc-1/calicoctl/kc1-calicoctl/bgp-peer/data-network-internet1-1]", // string
+      "type": "Modified", // "Added" | "Deleted" | "Modified"
+      "attributes": [
+        {
+          "id": "purged", // string
+          "type": "Modified", // "Modified" | "Unmodified"
+          "diff": "--- \n+++ \n@@ -0,0 +1 @@\n+false", // unified diff
+          "source": false, // old value
+          "target": false // new value
```
[the remark this answers is not in the document]
Comment (Stijnkool) [reply docs/desired-state-diff/index.md:122]: If we decide to do diffing in the frontend, we don't need the diff at all in the backend data.
So yea we need to discuss and choose a frontend option. And based on that we can decide on a data format.
But if the backend just decides to provide it all, that will always work.

File: docs/desired-state-diff/index.md [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+If you select the already select version for compare, you get an unselect option.
+
+![Unselect](./compare-5-unselect.png)
+
+## Libraries
+
+- jsdiff
+  - https://github.com/kpdecker/jsdiff
+  - Text diffing in js
+- react-diff-viewer
```
[the remark this answers is not in the document]
Comment (Stijnkool) [reply docs/desired-state-diff/index.md:35]: I think building it ourselves is quite doable if we use `jsdiff`.
I have added a quickly example to storybook where I diff 2 strings.
Syntax highlighting might be a different beast.

File: docs/desired-state-diff/index.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (bartv) [reply docs/desired-state-diff/index.md:122]: if the diff works well in the frontend, I would do it there. It provides more flexibility to do things like syntax high lighting

**Comment by bartv** [comment]
This library (perl) seems to generate the fancy diffs: https://github.com/so-fancy/diff-so-fancy where also the specific changes are highlighted.

**Comment by bartv** [comment]
Are there any other libraries than diff2html.js that can show the "fancy" diff? Like so

image

**Comment by Stijnkool** [comment]
What do you mean by "fancy"?
- Syntax highlighted
- line by line view (instead of side by side)
- line numbers
- hiding of unchanged lines?
- unified diff header
- file diff summary (+2 -3)

What are you looking for?
The view you shared is nice for files.
But we also have a lot of single line values where this diff view is less ideal.

File: src/UI/Pages/DesiredState/Components/CompareAction.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (andraskvr) [reply src/UI/Pages/DesiredState/Components/CompareAction.tsx:79]: I haven't checked the font awesome icons. If they don't have anything similar to this, you can disregard this comment

File: docs/desired-state-diff/index.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (andraskvr) [reply docs/desired-state-diff/index.md:35]: Yeah I meant mostly if we don't want to do the visuals ourselves.
If that's not a difficult part that's also fine by me

File: docs/desired-state-diff/index.md [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+
+## Selection
+
+How does the user select 2 versions to compare?
+
+This is a compare selection box next to the filters above all the versions.
+
+![Compare Initial](./compare-1-initial.png)
+
+On a version, click the actions dropdown and click "Select version for compare".
```
[the remark this answers is not in the document]
Comment (Stijnkool) [reply docs/desired-state-diff/index.md:13]: That makes sense to me. I will make the changes.

File: src/UI/Pages/DesiredState/Components/CompareAction.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (Stijnkool) [reply src/UI/Pages/DesiredState/Components/CompareAction.tsx:79]: I haven't found anything similar.

File: docs/desired-state-diff/index.md (hunk 1 above)
Comment (Stijnkool) [reply docs/desired-state-diff/index.md:202]: I think react-diff-viewer might be good enough. It's not really well maintained.
But as far as I can see, its got everything covered.

**Comment by Stijnkool** [comment]
The documentation changes were merged in with PR https://github.com/inmanta/web-console/pull/2358.
This branch still contains src changes that can be used later.

**Pull request closed without merging**